From 5e00f17fc72f94c9f9050955c13afa13c157550a Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Fri, 25 Sep 2020 19:03:07 +0200
Subject: [PATCH 650/985] evl/thread: align trap_entry/exit on new Dovetail
 rules

The trap_entry handler may expect hard irqs off on entry. It may
demote the current context to the in-band stage, may return with hard
irqs on.

The trap_exit handler only runs if trap_entry was called for the same
trap occurrence. It may expect hard irqs off on entry, may switch the
current context back to the oob stage. Must return with hard irqs
off. This departs from the previous implementation which allowed the
trap_exit handler to return with hard irqs on.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/thread.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/kernel/evl/thread.c b/kernel/evl/thread.c
index b58295c51..a2512db9e 100644
--- a/kernel/evl/thread.c
+++ b/kernel/evl/thread.c
@@ -1510,12 +1510,14 @@ static inline void note_trap(struct evl_thread *curr,
 			(void *)instruction_pointer(regs));
 }
 
-/* oob stalled. */
+/* hard irqs off. */
 void handle_oob_trap_entry(unsigned int trapnr, struct pt_regs *regs)
 {
 	struct evl_thread *curr;
 	bool is_bp = false;
 
+	trace_evl_thread_fault(trapnr, regs);
+
 	curr = evl_current();
 	if (curr->local_info & T_INFAULT) {
 		note_trap(curr, trapnr, regs, "recursive fault");
@@ -1526,8 +1528,6 @@ void handle_oob_trap_entry(unsigned int trapnr, struct pt_regs *regs)
 
 	curr->local_info |= T_INFAULT;
 
-	trace_evl_thread_fault(trapnr, regs);
-
 	if (current->ptrace & PT_PTRACED)
 		is_bp = evl_is_breakpoint(trapnr);
 
@@ -1541,11 +1541,13 @@ void handle_oob_trap_entry(unsigned int trapnr, struct pt_regs *regs)
 	evl_switch_inband(is_bp ? EVL_HMDIAG_TRAP : EVL_HMDIAG_EXDEMOTE);
 }
 
-/* hard irqs on. */
+/* hard irqs off */
 void handle_oob_trap_exit(unsigned int trapnr, struct pt_regs *regs)
 {
 	struct evl_thread *curr = evl_current();
 
+	hard_local_irq_enable();
+
 	curr->local_info &= ~T_INFAULT;
 
 	/*
@@ -1561,6 +1563,8 @@ void handle_oob_trap_exit(unsigned int trapnr, struct pt_regs *regs)
 		evl_switch_oob();
 		note_trap(curr, trapnr, regs, "resuming out-of-band");
 	}
+
+	hard_local_irq_disable();
 }
 
 void handle_oob_mayday(struct pt_regs *regs)
-- 
2.30.2

